### ISSUE TITLE
[BUG] sporadic dracut failures during baremetal deployment

### ISSUE THREAD

**Issue by guangyee** [harvester/harvester#1363]
**Describe the bug**
We are seeing `dracut-initqueue` timed out during baremetal deployment via PXE boot. The problem only occur sporadically. Subsequent re-deploy works. Also, the failure is not specific to a particular baremetal node as it has happened to different nodes.

**To Reproduce**
*NOTE:* the problem is not consistently reproducible as it only happen sporadically.
1. deploy a Harvester node via PXE boot.

**Expected behavior**
Deploy successfully.

**Support bundle**
None

**Environment:**
 - Harvester ISO version: master
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): Baremetal with HPE DL160

**Additional context**
![screenshot_from_2021-10-06_11-20-42](https://user-images.githubusercontent.com/1808417/136265196-ac0fb96d-4c1d-4cab-86f9-e37ccf25e4c0.png)

**Comment by bk201** [comment]
The sos log: [rdsosreport.txt](https://github.com/harvester/harvester/files/7308236/rdsosreport.txt)
Dracut fails to wait for the root device because squashfs image is not fetched to the system.

The DHCP process looks fine, but `livenetroot.sh` is never executed.
```
[   15.744508] localhost dracut-initqueue[765]: Starting dhcp for interface ens1f1
[   15.767788] localhost dracut-initqueue[841]: wicked: ens1f1: Request to acquire DHCPv4 lease with UUID 998b5f61-690f-0600-4903-000001000000
[   15.795930] localhost wicked[841]: ens1f1: Request to acquire DHCPv4 lease with UUID 998b5f61-690f-0600-4903-000001000000
[   15.810207] localhost kernel: NET: Registered protocol family 17
[   25.776590] localhost dracut-initqueue[841]: wicked: ens1f1: DHCP4 discovery failed
[   25.791442] localhost wicked[841]: ens1f1: DHCP4 discovery failed
[   25.799407] localhost dracut-initqueue[765]: Warning: DHCP failed
[   25.813718] localhost dracut-initqueue[765]: Warning: dhcp for interface ens1f1 failed
[   25.863977] localhost dracut-initqueue[765]: Starting dhcp for interface ens1f1
[   25.883117] localhost dracut-initqueue[884]: wicked: ens1f1: Request to acquire DHCPv4 lease with UUID a38b5f61-4ed2-0700-7403-000001000000
[   25.912364] localhost wicked[884]: ens1f1: Request to acquire DHCPv4 lease with UUID a38b5f61-4ed2-0700-7403-000001000000
[   35.888577] localhost dracut-initqueue[884]: wicked: ens1f1: DHCP4 discovery failed
[   35.903694] localhost wicked[884]: ens1f1: DHCP4 discovery failed
[   35.911260] localhost dracut-initqueue[765]: Warning: DHCP failed
[   35.925813] localhost dracut-initqueue[765]: Warning: dhcp for interface ens1f1 failed
[   37.090166] localhost kernel: bnx2x 0000:06:00.0 ens1f0: using MSI-X  IRQs: sp 64  fp[0] 66 ... fp[7] 75
[   37.416196] localhost kernel: bnx2i [06:00.00]: ISCSI_INIT passed
[   38.397967] localhost dracut-initqueue[895]: Starting dhcp for interface ens1f0
[   38.387674] localhost kernel: bnx2x 0000:06:00.0 ens1f0: NIC Link is Up, 10000 Mbps full duplex, Flow control: ON - receive & transmit
[   38.402663] localhost kernel: IPv6: ADDRCONF(NETDEV_CHANGE): ens1f0: link becomes ready
[   38.428922] localhost dracut-initqueue[971]: wicked: ens1f0: Request to acquire DHCPv4 lease with UUID b08b5f61-f5e3-0000-cb03-000001000000
[   38.460356] localhost wicked[971]: ens1f0: Request to acquire DHCPv4 lease with UUID b08b5f61-f5e3-0000-cb03-000001000000
[   48.433577] localhost dracut-initqueue[971]: wicked: ens1f0: DHCP4 discovery failed
[   48.450236] localhost wicked[971]: ens1f0: DHCP4 discovery failed
[   48.455570] localhost dracut-initqueue[895]: Warning: DHCP failed
[   48.455570] localhost dracut-initqueue[895]: Warning: dhcp for interface ens1f0 failed
[   48.524646] localhost dracut-initqueue[895]: Starting dhcp for interface ens1f0
[   48.548957] localhost dracut-initqueue[1012]: wicked: ens1f0: Request to acquire DHCPv4 lease with UUID ba8b5f61-bfb8-0200-f403-000001000000
[   48.581770] localhost wicked[1012]: ens1f0: Request to acquire DHCPv4 lease with UUID ba8b5f61-bfb8-0200-f403-000001000000
[   49.242456] localhost dracut-initqueue[1012]: wicked: ens1f0: Committed DHCPv4 lease with address 10.84.102.10 (lease time 21600 sec, renew in 10800 sec, rebind in 18900 sec)
[   49.276009] localhost wicked[1012]: ens1f0: Committed DHCPv4 lease with address 10.84.102.10 (lease time 21600 sec, renew in 10800 sec, rebind in 18900 sec)
[   49.297622] localhost dracut-initqueue[895]: DHCP is finished successfully
[  190.800952] localhost dracut-initqueue[623]: Warning: dracut-initqueue timeout - starting timeout scripts
[  191.428889] localhost dracut-initqueue[623]: Warning: dracut-initqueue timeout - starting timeout scripts
[  191.988441] localhost dracut-initqueue[623]: Warning: dracut-initqueue timeout - starting timeout scripts
```

@guangyee Could you try these two things:
- Use explicit interface name in boot parameters:  `ip=ens1f0:dhcp` in this case.
- Add the `rd.debug` parameter. This makes Dracuts output tons of debug information in sos reports.

**Comment by guangbochen** [comment]
@guangyee let me know if this issue is critical to u, I will temporarily move this to 1.1.0 for now, thanks

**Comment by alexdepalex** [comment]
We're running into this as well. And subsequently, for some reason, the nameservers we passed statically to the kernel through ipxe, are also not properly configured. Which results in the inability to resolve the download URL for the harvester config file.

**Comment by bk201** [comment]
@guangyee Is this still an issue after using a single interface like `ip=ens1f0:dhcp`? Thanks

**Comment by ibrokethecloud** [comment]
I am running into the same issue when trying to PXE boot baremetal nodes with tinkerbell.

Providing interface name is not a valid option at scale, since the interface name is not known before hand.

**Comment by ibrokethecloud** [comment]
I ended up with the following arguments to get a mac based lookup:

```ifname=bootnet:0c:c4:7a:6b:80:d0 ip=bootnet:dhcp```

**Comment by bk201** [comment]
@noahgildersleeve Do we still see this?

**Comment by johnliu55tw** [comment]
Based on my observation, when kernel parameter `ip=dhcp` is given (which is the case of this issue), dracut would request DHCP IP **twice** for each interface. However, if the first DHCP request failed and the second succeed, dracut **won't call the `livenet` module to retrieve the remote rootfs**. Only when the DHCP request succeeds for the **first try**, it fetches the rootfs.

A workaround is to increase the retry count using kernel parameter `rd.net.dhcp.retry=<cnt>`. `3` would be a good start.

**Comment by johnliu55tw** [comment]
Hi @alexdepalex are you still facing this issue? Or has the fix for https://github.com/harvester/harvester/issues/1590 solved it in your case?

**Comment by johnliu55tw** [comment]
I will update documents in https://github.com/harvester/ipxe-examples for the [workaround](https://github.com/harvester/harvester/issues/1363#issuecomment-1055149393).

**Comment by TachunLin** [comment]
Verified **fixed** on `master-7fc404f8-head` (05/06). Close this issue.

### Result
When user encounter **`failed to connect`** to DHCP server while PXE boot to acquire DHCP service. 
By adding the `rd.net.dhcp.retry=<cnt>` to the `imgargs` of pxe boot file, the installation process will automatically retry the given <cnt> retry to proceed the follow up instillation. 


* **With** `rd.net.dhcp.retry=<cnt>` added, pxe boot installation can retry and continue while the DHCP connection come back
  ```
  #!ipxe
  
  kernel http://192.168.0.254/harvester/harvester-vmlinuz-amd64
  initrd http://192.168.0.254/harvester/harvester-initrd-amd64
  imgargs harvester-vmlinuz-amd64 rd.net.dhcp.retry=5 initrd=harvester-initrd-amd64 ip=ens6:dhcp net.ifnames=1 rd.cos.disable rd.live.debug=1 rd.noverifyssl root=live:http://192.168.0.254/harvester/harvester-rootfs-amd64.squashfs console=tty1 harvester.install.automatic=true harvester.install.config_url=http://192.168.0.254/harvester/config-create.yaml
  boot
  ```
  ![image](https://user-images.githubusercontent.com/29251855/167571632-458f3b38-3c34-463c-a7e5-1a697d8d5f35.png)

* **Without** `rd.net.dhcp.retry=<cnt>`,  encounter dracut timeout failure
  ![image](https://user-images.githubusercontent.com/29251855/167569497-08039c35-9005-43f9-bf27-265f4f6575de.png)


### Test Information
* Test Environment: 1 node harvester on local kvm machine
* Harvester version: master-7fc404f8-head


### Verify Steps
1. Clone [ipxe-example](https://github.com/harvester/ipxe-examples) project
2. Edit settings.yaml, set only one node
3. Run ./setup_harvester.sh
4. Wait for the PXE boot server installation complete
5. Pause the harvester node 1 installation 
6. Run `ssh vagrant pxe_server` to login pxe server
7. Edit the boot file of the node machine MAC address
  ```
  root@pxe-server:/var/www/harvester# ls
  02:00:00:0d:62:e2  config-create.yaml  harvester-amd64.iso     harvester-rootfs-amd64.squashfs
  02:00:00:35:86:92  config-join-1.yaml  harvester-initrd-amd64  harvester-vmlinuz-amd64
  
  root@pxe-server:/var/www/harvester# vim 02:00:00:0d:62:e2
  ```
8. Change the `imgargs` from `ip=ens5:dhcp` to `ip=ens6:dhcp` 
9. Change the server url from from `http://192.168.121.212` to `http://192.168.0.254`
  ```
  #!ipxe
  
  kernel http://192.168.0.254/harvester/harvester-vmlinuz-amd64
  initrd http://192.168.0.254/harvester/harvester-initrd-amd64
  imgargs harvester-vmlinuz-amd64 initrd=harvester-initrd-amd64 ip=ens6:dhcp net.ifnames=1 rd.cos.disable rd.live.debug=1 rd.noverifyssl root=live:http://192.168.0.254/harvester/harvester-rootfs-amd64.squashfs console=tty1 harvester.install.automatic=true harvester.install.config_url=http://192.168.0.254/harvester/config-create.yaml
  boot
  ```
10. Continue the node 1 installation
11. On node 1 console, once node 1 PXE boot acquire DHCP service can get the initrd and rootfs 
12. On pxe server, run `systemctl stop isc-dhcp-server.service` to stop the DHCP service
13. Wait for 10 minutes, then the console will show dracut timeout failure
  ![image](https://user-images.githubusercontent.com/29251855/167569497-08039c35-9005-43f9-bf27-265f4f6575de.png)
  ![image](https://user-images.githubusercontent.com/29251855/167570207-342da94d-93d1-4793-bf83-067756aa0d51.png)
14. Edit the boot file of the node machine MAC address
15. Add the `rd.net.dhcp.retry=5` to  `imgargs`
  ```
  #!ipxe
  
  kernel http://192.168.0.254/harvester/harvester-vmlinuz-amd64
  initrd http://192.168.0.254/harvester/harvester-initrd-amd64
  imgargs harvester-vmlinuz-amd64 rd.net.dhcp.retry=3 initrd=harvester-initrd-amd64 ip=ens6:dhcp net.ifnames=1 rd.cos.disable rd.live.debug=1 rd.noverifyssl root=live:http://192.168.0.254/harvester/harvester-rootfs-amd64.squashfs console=tty1 harvester.install.automatic=true harvester.install.config_url=http://192.168.0.254/harvester/config-create.yaml
  boot
  ```
16. Reboot the node 1 VM 
17. Console will display failed to acquire DHCP service 
18. On pxe server, run `systemctl start isc-dhcp-server.service` to start DCHP service
19. Check installation can retry and successfully acquire DCHP server to proceed the process.
